pytorch_lightning/utilities/auto_restart.py

```diff
@@ -207,29 +210,195 @@ def __iter__(self) -> Iterator:
+
+    @staticmethod
+    def _sanitize_batch_from_sampler_state(data: Any, state_dicts: List):
+        """
+        This function is used to remove the sampler state dict from provided data batch.
+        The custom data has this format:
+
+        .. code-block:: python
+
+            {
+                "batch": data returned by DataLoader
+                "__pl_samplers": {
+                    "sampler0": {
+                        0: { "current_iteration": ... }
+                        1: { "current_iteration": ... }
+                        ...
+                    }
+                    "sampler1": ...
+                }
+            }
+
+        Each sampler in the worker process tracks the current iteration. We return all of them to the main process
+        as part of the sample and then a special collate function :func:`_sampler_metadata_collate`
+        will extract the current iteration as part of the metadata returned by a custom batch.
+        """
+
+        def _sanitize(data: Mapping):
+            out = []
+            for k, v in data.items():
+                if k == AutoRestartBatchKeys.PL_SAMPLERS:
+                    state_dicts.append(v)
+                    return data["data"]
+                out.append((k, CaptureIterableDataset._sanitize_batch_from_sampler_state(v, state_dicts)))
+            return out
 
-        # return both current data and samplers ``state_dict``.
-        return {"data": data, AutoRestartBatchKeys.PL_SAMPLERS: state_dicts}
+        return apply_to_collection(data, Mapping, _sanitize)
 
     @staticmethod
-    def convert_batch_into_state_dict(batch) -> Dict[str, Dict[int, Any]]:
+    def extract_samplers_state_dict_from_batch(batch) -> List[Dict[int, Any]]:
         """
         This function is used to convert a batch into a state_dict
         """
-        state_dict = {}
-        batch_worker_id = batch[AutoRestartBatchKeys.PL_SAMPLERS].pop("id")
-        worker_id = batch_worker_id[-1].item()
-        for sampler_name, sampler_state_dict in batch[AutoRestartBatchKeys.PL_SAMPLERS].items():
-            state_dict[sampler_name] = {
-                worker_id: {
-                    "current_iteration": sampler_state_dict[worker_id]["current_iteration"][-1].item()
-                }
+        samplers_state_dict = []
+
+        batch = CaptureIterableDataset._sanitize_batch_from_sampler_state(batch, samplers_state_dict)
+
+        return batch, samplers_state_dict
+
+
+def _find_fast_forward_samplers(dataloader: DataLoader) -> Optional[FastForwardSampler]:
+    """
+    If the ``DataLoader`` is wrapping a mapping based Dataset, return the ``FastForwardSampler``.
+    """
+    if isinstance(dataloader.sampler, FastForwardSampler):
+        return dataloader.sampler
+
+    if isinstance(dataloader.batch_sampler, FastForwardSampler):
+        return dataloader.batch_sampler
+
+
+def _cycle_to_next_worker_and_reset(dataloader: DataLoader, state_dict: Dict[str, Any]) -> Iterator:
```

Comment:
I see plenty of codecov annotation, do we have this tested somewhere?

Reply:
Yes, but we don't get coverage within workers.